Use/enable TMC2130 DEDGE support in MK3/MK3S

wavexx opened the pull request:
The TMC2130 support Double Edge stepping aka ``DEDGE`` where both the rising and lowering edge of the stepping pin count as a step instead of requiring a full cycle.

DEDGE is advantageous as it removes one extra toggle for each stepper pulsed (plus a delay, in some circumstances), saving a few µs here and there.

We introduce 2 macros to step: STEP_NC_HI and STEP_NC_LO.

In normal mode, both perform as their name implies. In DEDGE mode instead STEP_NC_HI inverts the line, while STEP_NC_LO becomes a no-op. Both modes are still fully supported. The macros are specialized for each axis, and still respect INVERT_?_AXIS as defined in the variant file.

We also introduce a few constant definitions the delays required by the TMC2130, and use them when appropriate instead of hard-coding delays:

- step pauses (2µs minimum)
- direction changes (100µs minimum)

Existing code that was manipulating the pins directly has been changed to use either the macros or the helper functions defined in tmc2130.h. We also fix the behavior of TMC2130_INTPOL_* which was previously ignored in the variant files.

PFW-1192

wavexx committed:
Reintroduce the ability to disable TMC interpolation per-axis

wavexx committed:
Enable DEDGE stepping on supported variants

wavexx committed:
Implement proper step/delay pauses in tmc2130 functions

Add constants for the various required delays in tmc2130.h,
which will come in handy for stepper.cpp as well.

Move the delays in the _set functions and remove the pauses
from the various calling points and macros.

Note that the hard-coded pause wouldn't cut it for the stepper ISR,
but it's fine for other use cases.

wavexx committed:
Fix DEDGE in sm4.c (fixes xyz calibration)

wavexx committed:
Fix delay calculations inside babystep()

- Avoid all delays when using DEDGE stepping
- Correctly account for direction change delays

wavexx committed:
Implement double-edge stepping

Introduce new wrapper macros to tick the stepper pins.
Default to the original raising-edge stepping mode.

When using the TMC double-edge stepping mode (aka half-wave or
square-wave mode) the _LO macros become no-ops.

DRracer commented:
@wavexx @vintagepc is this ready to get merged into MK3?

DRracer commented on Firmware/tmc2130.cpp:
```diff
@@ -678,25 +696,32 @@ static uint8_t tmc2130_rx(uint8_t axis, uint8_t addr, uint32_t* rval)
+#define _DO_STEP_Z      TOGGLE(Z_STEP_PIN)
+#define _DO_STEP_E      TOGGLE(E0_STEP_PIN)
+#else
+#define _DO_STEP_X      { WRITE(X_STEP_PIN, !INVERT_X_STEP_PIN); delayMicroseconds(TMC2130_MINIMUM_PULSE); WRITE(X_STEP_PIN, INVERT_X_STEP_PIN); }
```
calling/inlining delayMicroseconds will definitely take more time than the original "nop" instruction, why was this changed?

vintagepc replied:
Looks like it's to support variable minimum pulse times. Could probably special-case it in a #define so that if the pulse time is 0 it just `nop`s but otherwise adds a delay call

leptun replied:
Do we really need that nop? It’s not like it makes any real difference

wavexx replied:
I can make an exception for TMC2130 to use the nop as before, much nicer.

_DO_STEP* is not used by the stepping code, this is essentially just for XYZCAL. But removing the the call is still nice ;)

vintagepc replied:
> Do we really need that nop? It’s not like it makes any real difference

Preventing Compiler optimizations that could eliminate the actual change, perhaps?

wavexx replied:
Note that some delay is necessary by the spec. A nop seems to be always sufficient on the Einsy, I wouldn't remove it completely. On an externally wired 2130 the TMC2130_MINIMUM_PULSE is fully required.

leptun replied:
It is not optimized away. I've done 8MHz toggles before. I'm pretty sure the registers are defined as volatile.
As for external drivers, never had issues with missed step pulses, even with 20cm wires and horrible board design

wavexx replied:
Ok, I'll remove the nop then. I actually ran into missed step though in my case

wavexx replied:
(on an external 2130)

wavexx replied:
Ok, I removed all delays entirely. @leptun could you try a build _without_ DEDGE and see if everything is still a-ok?

There are still a couple of "nops" in sm4.c for XYZCAL.
We could go the extra mile and remove those as well.

leptun replied:
I wanted do say that there should be no NOP when DEDGE is used. With regular stepping it is unfortunately needed as you've said. I thought you meant to keep the delay with DEDGE, which didn't make sense to me. There should only be delays between the HIGH->LOW transition of the nonDEDGE mode

wavexx replied:
I'll refine this further in a few minutes

wavexx replied:
Ok, the regular variant now restores the nop instruction, but still allows to use delayMicroseconds when not using TMC2130. DEDGE removes all details (as it was the case before).

wavexx committed:
Elide delayMicroseconds for TMC2130 in non-DEDGE mode

Introduce new macros TMC2130_MINIMUM_DELAY/STEPPER_MINIMUM_DELAY for
blocking pauses.

If MINIMUM_PULSE has defined to be zero, avoid the delay call entirely.

wavexx committed:
Remove TMC2130 special-cases

With the new STEPPER_MINIMUM_DELAY being automatically removed for
TMC2130 we no longer need to add specialized #ifdefs for DEDGE in
babystep.

wavexx committed:
Ensure MINIMUM_PULSE is always 0 in DEDGE mode

This ensures delays are always properly elided without having to check
for DEDGE all over the place.

wavexx committed:
Also fix delay instances inside unused BACKLASH_[XY]

wavexx committed:
Reinstate the nop instruction as delay in non-DEDGE

When TMC2130_MINIMUM_PULSE is 0 a minimum delay is implied.
In this case, use a single "nop" instruction.

wavexx committed:
Also toggle pins efficiently in sm4.c

Use the same technique used in fastio to toggle pins efficiently in sm4
when DEDGE is used.

wavexx commented:
Hopefully I've ironed out all concerns. The regular (non-DEDGE) variant now uses a single nop instruction as it used to, but still allows variable delays if needed.

Tweaked sm4.cpp to use the same efficient toggle mechanism as used by stepper.cpp.

I've tested both variants merged on the current master and both work as expected, including XYZcal.

DRracer commented on Firmware/tmc2130.h:
```diff
@@ -29,6 +29,18 @@ extern uint8_t tmc2130_sg_homing_axes_mask;
+#define TMC2130_SET_PWR_DELAY 0   // minimum delay after changing pwr mode in uS
+
+#ifdef TMC2130_DEDGE_STEPPING
+#define TMC2130_MINIMUM_DELAY //NOP
```
Personally I prefer `/*NOP*/`, because that terminates the comment when placed into the source code by the preprocessor

wavexx replied:
Isn't the comment already eaten during parsing? (the expression should expand to nothing)

wavexx replied:
```
#define A b //NOP
A C
```
```
% gcc -E test.c
# 1 "test.c"
# 1 "<built-in>"
# 1 "<command-line>"
# 31 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 32 "<command-line>" 2
# 1 "test.c"

b C
```

Looks like (I was going from memory)

DRracer replied:
ok
